hivemind/moe/client/expert.py

```diff
@@ -52,18 +81,171 @@ def forward(self, *args, **kwargs):
+
+    @classmethod
+    def run_coroutine(cls, coro: Awaitable, return_future: bool = False):
+        if cls._event_thread is None or cls._pid != os.getpid():
+            cls._pid = os.getpid()
+            cls._event_thread = Thread(target=cls._run, daemon=True)
+            cls._event_thread.start()
+
+        future = Future()
+        cls._task_queue.put((coro, future))
+
+        if return_future:
+            return future
+
+        result = future.result()
+        return result
+
+    @classmethod
+    def _spawn_experts(cls, infos: Sequence[Optional[RemoteExpertInfo]], p2p: P2P) -> List[Optional[RemoteExpert]]:
+        experts: List[Optional[RemoteExpert]] = []
+        for i in infos:
+            if i is not None:
+                experts.append(RemoteExpert(i, p2p))
+            else:
+                experts.append(None)
+        return experts
+
+    @classmethod
+    def spawn_experts(
+        cls, infos: Union[Sequence[Optional[RemoteExpertInfo]], MPFuture], dht: DHT, return_future: bool = False
+    ) -> Union[List[Optional[RemoteExpert]], Future]:
+        if return_future:
+
+            async def _unpack(infos_future: MPFuture, dht: DHT):
+                p2p = await dht.replicate_p2p()
+                return cls._spawn_experts(await infos_future, p2p)
+
+            return cls.run_coroutine(_unpack(infos, dht), return_future)
+
+        p2p = cls.run_coroutine(dht.replicate_p2p())
+        return cls._spawn_experts(infos, p2p)
+
+    @classmethod
+    def batch_spawn_experts(
+        cls,
+        infos: Union[Sequence[Sequence[Optional[RemoteExpertInfo]]], MPFuture],
+        dht: DHT,
+        return_future: bool = False,
+    ) -> Union[List[List[Optional[RemoteExpert]]], Future]:
+        if return_future:
+
+            async def _unpack(infos_future: MPFuture, dht: DHT):
+                p2p = await dht.replicate_p2p()
+                return [cls._spawn_experts(i, p2p) for i in await infos_future]
+
+            return cls.run_coroutine(_unpack(infos, dht), return_future)
+
+        return [cls.spawn_experts(exps, dht) for exps in infos]
+
+
+async def _backward_stream(uid: str, serialized_tensors: Iterable[runtime_pb2.Tensor], stub) -> List[torch.Tensor]:
+    split = (p for t in serialized_tensors for p in split_for_streaming(t, DEFAULT_MAX_MSG_SIZE // 2))
+
+    grad_inputs = await stub.rpc_backward_stream(
+        amap_in_executor(
+            lambda tensor: runtime_pb2.ExpertRequest(uid=uid, tensors=[tensor]),
+            iter_as_aiter(split),
+        ),
+    )
+
+    return await gather_from_streaming(grad_inputs, lambda r: r.tensors, deserialize_torch_tensor)
+
+
+async def _backward_unary(uid: str, serialized_tensors: Iterable[runtime_pb2.Tensor], stub) -> List[torch.Tensor]:
+    grad_inputs: runtime_pb2.ExpertResponse = await stub.rpc_backward(
+        runtime_pb2.ExpertRequest(uid=uid, tensors=list(serialized_tensors))
+    )
+    return [deserialize_torch_tensor(t) for t in grad_inputs.tensors]
```

Comment:
Side note: it's generally best to avoid CPU-heavy work inside asyncio, since it stalls the event loop. Thus, we need to be extremely careful with such constructions and place them in separate processes: otherwise, we lose nearly all the profit from asynchronous execution

Reply:
Agreed, but I would do this changes later, because it seems to be a thing which do not change functionality. It is rather performance tune, and performance right now is acceptable for merging.

Added that into list of thing to be done afterwards.